File: JANL.VB/SQL/SelectProcedure.vb
Imports System.Data.SqlClient
Imports System.Runtime.CompilerServices
Imports System.Threading.Tasks

Namespace SQL
	''' <summary>
	''' Процедура возвращающая <see cref="DataTable"/>
	''' </summary>
	Public Class SelectProcedure
		Inherits SQLProcedure(Of DataTable)

		''' <summary>
		''' Создаёт новую процедуру с именем вызывающего метода
		''' </summary>
		Public Sub New(<CallerMemberName> Optional Name As String = Nothing)
			MyBase.New(Name)
		End Sub

		''' <summary>
		''' Выполнить с указанным соединением
		''' </summary>
		Public Overrides Function Execute(Connection As SqlConnection) As DataTable
			Dim Result = New DataTable() With {.Locale = Globalization.CultureInfo.CurrentCulture}
			SQLCommand.Connection = Connection
			Using adapter As New SqlDataAdapter(SQLCommand)
				adapter.Fill(Result)
			End Using
			Return Result
		End Function

		''' <summary>
		''' Асинхронно выполнить с указанным соединением
		''' </summary>
		Public Overrides Async Function ExecuteAsync(Connection As SqlConnection) As Task(Of DataTable)
			Dim Result = New DataTable() With {.Locale = Globalization.CultureInfo.CurrentCulture}
			SQLCommand.Connection = Connection
			Using R = Await SQLCommand.ExecuteReaderAsync()
				Await Task.Run(Sub() Result.Load(R))
			End Using
			Return Result
		End Function

	End Class
End Namespace

File: JANL/SQL/SelectProcedure.cs
using System.Data;
using System.Data.SqlClient;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

namespace JANL.SQL
{
    /// <summary>
    /// Процедура возвращающая <see cref="DataTable"/>
    /// </summary>
    public class SelectProcedure : SQLProcedure<DataTable>
    {
        /// <summary>
        /// Создаёт новую процедуру именем вызывающего метода
        /// </summary>
        public SelectProcedure([CallerMemberName] string Name = null) : base(Name) { }

        /// <summary>
        /// Выполнить с указанным соединением
        /// </summary>
        public override DataTable Execute(SqlConnection Connection)
        {
            var Result = new DataTable { Locale = CultureInfo.CurrentCulture };
            Command.Connection = Connection;
            //TODO Replace with Reader (Blocked by RGI)
            using (var A = new SqlDataAdapter(Command))
            {
                A.Fill(Result);
            }
            return Result;
        }

        /// <summary>
        /// Асинхронно выполнить с указанным соединением
        /// </summary>
        public override async Task<DataTable> ExecuteAsync(SqlConnection Connection)
        {
            var Result = new DataTable { Locale = CultureInfo.CurrentCulture };
            Command.Connection = Connection;
            using (var R = await Command.ExecuteReaderAsync().ConfigureAwait(false))
            {
                await Task.Run(() => Result.Load(R)).ConfigureAwait(false);
            }
            return Result;
        }

        /// <summary>
        /// Создать Reader
        /// </summary>
        public SqlDataReader ExecuteReader()
        {
            var Connection = new SqlConnection(ConnectionString);
            Connection.Open();
            Command.Connection = Connection;
            return Command.ExecuteReader(CommandBehavior.CloseConnection);
        }

        /// <summary>
        /// Создать Reader
        /// </summary>
        public async Task<SqlDataReader> ExecuteReaderAsync()
        {
            var Connection = new SqlConnection(ConnectionString);
            await Connection.OpenAsync();
            Command.Connection = Connection;
            return await Command.ExecuteReaderAsync(CommandBehavior.CloseConnection);
        }
    }
}